### PR TITLE
python-api: cleanup, better serial support

### PR DESCRIPTION
**Pull request by mlyle**
- refactor serial objects to directly use pyserial primitives.  This is
progress towards the API working well with hardware on windows.
- use enums inside the handshaking routine in the telemetry class, along
with defaults.
- adjust telemetry class to initially send "disconnected" state
- remove debugging try/except inside uavo.py
- use the uavo_collection to find parents instead of accessing global
namespace
- make updating global namespace optional in the innermost parts of
python-api -- this will be plumbed out in the future
- fix find_by_name() call to support current uavo_collection dictionary
semantics
- add a get_config.py that retrieves configuration from a flight board

### CODE REVIEW

**Comment by mlyle** [comment]
This is work-in-progress towards automated upgrade / uavo adaptation.. as a nicety for developers and a proof-of-concept that we can do much better in this area.  I figured it has reached a good first point for review.

**Comment by mlyle** [comment]
This is output from the dump configuration script on a random sparky2:

```
Michaels-MBP:python mlyle$ ./get_config.py -s -b 115200 /dev/tty.usbmodem1451 
Disconnected
Disconnected
Handshake ackd
Handshake ackd
Connected
Connected
Connected
UAVO_ActuatorSettings(name=UAVO_ActuatorSettings, time=1445021321.75, uavo_id=0x3557036A, MotorInputOutputCurveFit=(1.0, 1.0), TimerUpdateFreq=(50, 50, 50, 50, 50, 50), ChannelMax=(2000, 2000, 2000, 2000, 2000, 2000, 2000, 2000, 2000, 2000), ChannelNeutral=(1000, 1000, 1000, 1000, 1000, 1000, 1000, 1000, 1000, 1000), ChannelMin=(1000, 1000, 1000, 1000, 1000, 1000, 1000, 1000, 1000, 1000), TimerPwmResolution=(1MHz(0), 1MHz(0), 1MHz(0), 1MHz(0), 1MHz(0), 1MHz(0)), ChannelType=(PWM(0), PWM(0), PWM(0), PWM(0), PWM(0), PWM(0), PWM(0), PWM(0), PWM(0), PWM(0)), MotorsSpinWhileArmed=FALSE(0))
UAVO_AttitudeSettings(name=UAVO_AttitudeSettings, time=1445021321.79, uavo_id=0x9DB8D0A, MagKp=0.0500000007451, MagKi=9.99999974738e-05, AccelKp=0.0500000007451, AccelKi=9.99999974738e-05, AccelTau=0.10000000149, VertPositionTau=2.0, YawBiasRate=9.99999997475e-07, BoardRotation=(0, 0, 0), ZeroDuringArming=TRUE(1), BiasCorrectGyro=TRUE(1), FilterChoice=CCC(0))
UAVO_HomeLocation(name=UAVO_HomeLocation, time=1445021322.79, uavo_id=0xCA32B032, Latitude=0, Longitude=0, Altitude=0.0, Be=(0.0, 0.0, 0.0), GroundTemperature=150, SeaLevelPressure=1013, Set=FALSE(0))
UAVO_HwSparky2(name=UAVO_HwSparky2, time=1445021323.79, uavo_id=0x5E852A4C, CoordID=0, RcvrPort=Disabled(0), MainPort=Disabled(0), FlexiPort=Disabled(0), USB_HIDPort=Disabled(2), USB_VCPPort=USBTelemetry(0), DSMxMode=Autodetect(0), Radio=Disabled(0), MaxRfSpeed=64000(3), MaxRfPower=1.25(1), MinChannel=0, MaxChannel=250, GyroRange=2000(3), AccelRange=8G(2), MPU9250Rate=500(3), MPU9250GyroLPF=184(0), MPU9250AccelLPF=184(1), VTX_Ch=1(0), Magnetometer=Internal(0), ExtMagOrientation=Top0degCW(0))
UAVO_INSSettings(name=UAVO_INSSettings, time=1445021324.06, uavo_id=0xE2DBC224, AccelVar=(0.003000000026077032, 0.003000000026077032, 0.003000000026077032), GyroVar=(9.999999747378752e-06, 9.999999747378752e-06, 9.999999747378752e-05), MagVar=(10.0, 10.0, 100.0), GpsVar=(0.0010000000474974513, 0.009999999776482582, 0.5), BaroVar=0.00999999977648, MagBiasNullingRate=0.0, ComputeGyroBias=FALSE(0))
UAVO_MWRateSettings(name=UAVO_MWRateSettings, time=1445021324.09, uavo_id=0xBD3B5F28, RollRatePID=(0.004000000189989805, 0.00860000029206276, 4.999999873689376e-05, 0.30000001192092896), PitchRatePID=(0.004000000189989805, 0.00860000029206276, 4.999999873689376e-05, 0.30000001192092896), YawRatePID=(0.008500000461935997, 0.013000000268220901, 0.0, 0.30000001192092896), DerivativeGamma=0.0, RollPitchRate=0, YawRate=0)
UAVO_ManualControlSettings(name=UAVO_ManualControlSettings, time=1445021324.11, uavo_id=0xCC4A79BC, Deadband=0.0, RssiMax=2000, RssiMin=1000, ChannelMin=(1000, 1000, 1000, 1000, 1000, 1000, 1000, 1000, 1000, 1000), ChannelNeutral=(1500, 1500, 1500, 1500, 1500, 1500, 1500, 1500, 1500, 1500), ChannelMax=(2000, 2000, 2000, 2000, 2000, 2000, 2000, 2000, 2000, 2000), ArmedTimeout=30000, ChannelGroups=(None(8), None(8), None(8), None(8), None(8), None(8), None(8), None(8), None(8), None(8)), RssiType=None(0), RssiChannelNumber=0, ChannelNumber=(0, 0, 0, 0, 0, 0, 0, 0, 0, 0), Arming=Always Disarmed(0), ArmTime=1000(2), DisarmTime=1000(2), Stabilization1Settings=(Attitude(2), Attitude(2), Rate(1)), Stabilization2Settings=(Attitude(2), Attitude(2), Rate(1)), Stabilization3Settings=(Attitude(2), Attitude(2), Rate(1)), FlightModeNumber=3, FlightModePosition=(Manual(0), Stabilized1(7), Stabilized2(8), Stabilized3(9), ReturnToHome(13), PositionHold(12)), ArmTimeoutAutonomous=ENABLED(1))
UAVO_MixerSettings(name=UAVO_MixerSettings, time=1445021324.12, uavo_id=0x4E485B94, ThrottleCurve1=(0.0, 0.25, 0.5, 0.75, 1.0), ThrottleCurve2=(0.0, 0.25, 0.5, 0.75, 1.0), Curve2Source=Throttle(0), Mixer1Type=Motor(1), Mixer1Vector=(127, 0, 63, 63, -63), Mixer2Type=Motor(1), Mixer2Vector=(127, 0, -63, 63, 63), Mixer3Type=Motor(1), Mixer3Vector=(127, 0, -63, -63, -63), Mixer4Type=Motor(1), Mixer4Vector=(127, 0, 63, -63, 63), Mixer5Type=Disabled(0), Mixer5Vector=(0, 0, 0, 0, 0), Mixer6Type=Disabled(0), Mixer6Vector=(0, 0, 0, 0, 0), Mixer7Type=Disabled(0), Mixer7Vector=(0, 0, 0, 0, 0), Mixer8Type=Disabled(0), Mixer8Vector=(0, 0, 0, 0, 0), Mixer9Type=Disabled(0), Mixer9Vector=(0, 0, 0, 0, 0), Mixer10Type=Disabled(0), Mixer10Vector=(0, 0, 0, 0, 0))
UAVO_ModuleSettings(name=UAVO_ModuleSettings, time=1445021324.14, uavo_id=0x7D5CE5BA, AdminState=(Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0), Disabled(0)), TelemetrySpeed=57600(5), GPSSpeed=57600(5), GPSDataProtocol=UBX(1), GPSAutoConfigure=TRUE(1), GPSConstellation=GPS(1), GPSSBASConstellation=All(0), GPSDynamicsMode=Airborne2G(4), ComUsbBridgeSpeed=57600(5), I2CVMProgramSelect=None(2), MavlinkSpeed=57600(5), LightTelemetrySpeed=2400(1), FrskyAccelData=Accels(0))
UAVO_SensorSettings(name=UAVO_SensorSettings, time=1445021324.65, uavo_id=0xADB3B106, AccelBias=(0.059354428201913834, 0.11299746483564377, -0.01307495404034853), AccelScale=(0.9971875548362732, 0.9995233416557312, 0.990854799747467), GyroScale=(1.0, 1.0, 1.0), XGyroTempCoeff=(-0.34940001368522644, 0.0, 0.0, 0.0), YGyroTempCoeff=(-0.41538000106811523, 0.0, 0.0, 0.0), ZGyroTempCoeff=(0.1387300044298172, 0.0, 0.0, 0.0), MagBias=(-137.27850341796875, -285.8454284667969, -446.6678161621094), MagScale=(1.3035469055175781, 1.3113784790039062, 1.208706259727478), ZAccelOffset=0.0)
UAVO_StabilizationSettings(name=UAVO_StabilizationSettings, time=1445021324.65, uavo_id=0xCD238D1E, ManualRate=(150.0, 150.0, 150.0), MaximumRate=(300.0, 300.0, 300.0), PoiMaximumRate=(30.0, 30.0, 30.0), RollRatePID=(0.0020000000949949026, 0.001500000013038516, 0.0, 0.30000001192092896), PitchRatePID=(0.0020000000949949026, 0.001500000013038516, 0.0, 0.30000001192092896), YawRatePID=(0.0035000001080334187, 0.0035000001080334187, 0.0, 0.30000001192092896), RollPI=(2.5, 0.0, 50.0), PitchPI=(2.5, 0.0, 50.0), YawPI=(2.5, 0.0, 50.0), VbarSensitivity=(0.5, 0.5, 0.5), VbarRollPID=(0.004999999888241291, 0.0020000000949949026, 0.0), VbarPitchPID=(0.004999999888241291, 0.0020000000949949026, 0.0), VbarYawPID=(0.004999999888241291, 0.0020000000949949026, 0.0), VbarTau=0.5, GyroCutoff=55.0, DerivativeGamma=1.0, WeakLevelingKp=0.10000000149, CoordinatedFlightYawPI=(0.0, 0.10000000149011612, 0.5), AcroInsanityFactor=40.0, RollMax=55, PitchMax=55, YawMax=35, RateExpo=(0, 0, 0), AttitudeExpo=(0, 0, 0), HorizonExpo=(30, 30, 30), RollRateTPA=(75, 0), PitchRateTPA=(75, 0), YawRateTPA=(75, 0), VbarGyroSuppress=30, VbarPiroComp=FALSE(0), VbarMaxAngle=10, DerivativeCutoff=20, MaxAxisLock=15, MaxAxisLockRate=2, MaxWeakLevelingRate=5, LowThrottleZeroIntegral=TRUE(1))
UAVO_StateEstimation(name=UAVO_StateEstimation, time=1445021324.67, uavo_id=0xBA09F7C2, AttitudeFilter=Complementary(0), NavigationFilter=INS(2))
UAVO_SystemSettings(name=UAVO_SystemSettings, time=1445021324.71, uavo_id=0x6F3BB6B0, AirframeCategorySpecificConfiguration=(874577920, 0, 0, 0), AirframeType=QuadX(5))
UAVO_TrimAnglesSettings(name=UAVO_TrimAnglesSettings, time=1445021324.72, uavo_id=0x4E09A748, Roll=0.0, Pitch=0.0)
No instance of UAVO_AirspeedSettings
No instance of UAVO_AltitudeHoldSettings
No instance of UAVO_BrushlessGimbalSettings
No instance of UAVO_CameraStabSettings
No instance of UAVO_FaultSettings
No instance of UAVO_FixedWingAirspeeds
No instance of UAVO_FixedWingPathFollowerSettings
No instance of UAVO_FlightBatterySettings
No instance of UAVO_FlightPlanSettings
No instance of UAVO_FlightStatsSettings
No instance of UAVO_GeoFenceSettings
No instance of UAVO_GroundPathFollowerSettings
No instance of UAVO_HoTTSettings
No instance of UAVO_HwAQ32
No instance of UAVO_HwColibri
No instance of UAVO_HwCopterControl
No instance of UAVO_HwDiscoveryF4
No instance of UAVO_HwFlyingF3
No instance of UAVO_HwFlyingF4
No instance of UAVO_HwNaze
No instance of UAVO_HwQuanton
No instance of UAVO_HwRevoMini
No instance of UAVO_HwShared
No instance of UAVO_HwSparky
No instance of UAVO_HwSparkyBGC
No instance of UAVO_HwTauLink
No instance of UAVO_I2CVMUserProgram
No instance of UAVO_LoggingSettings
No instance of UAVO_OpenLRS
No instance of UAVO_OpticalFlowSettings
No instance of UAVO_OveroSyncSettings
No instance of UAVO_PathPlannerSettings
No instance of UAVO_PicoCSettings
No instance of UAVO_SystemIdent
No instance of UAVO_TxPIDSettings
No instance of UAVO_VibrationAnalysisSettings
No instance of UAVO_VtolPathFollowerSettings
```

File: python/get_config.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/python -B
+
+# Insert the parent directory into the module import search path.
+import os
+import sys
+sys.path.insert(1, os.path.dirname(sys.path[0]))
```
Comment (mluessi) [inline python/get_config.py]: This shouldn't be necessary, as you are already in the "python" directory and the current directory is always added to the path

File: python/get_config.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+import argparse
+import errno
+import code
+import struct
+import time
+from taulabs import uavo, telemetry, uavo_collection
```
Comment (mluessi) [inline python/get_config.py]: There should be a blank line before this one (separate standard modules from others)

File: python/taulabs/telemetry.py [hunk 3]
```diff
@@ -302,12 +317,46 @@ def _receive(self, finish_time):
 
         return ret
 
+    def _send(self, msg):
+        """ Send a string out the TCP socket """
```
Comment (mluessi) [inline python/taulabs/telemetry.py]: Doesn't have to be TCP

File: python/taulabs/telemetry.py [hunk 4]
```diff
@@ -302,12 +317,46 @@ def _receive(self, finish_time):
...
+
+        self.fd = fd
+
     # Call select and do one set of IO operations.
     def _do_io(self, finish_time):
         rdSet = []
         wrSet = []
 
-        didStuff = False
+        did_stuff = False
```
Comment (mluessi) [inline python/taulabs/telemetry.py]: Bye Bye CamelCase :)

File: python/taulabs/uavo.py [hunk 5]
```diff
@@ -231,19 +231,15 @@ def make_class(xml_file):
...
-                        for k, v in sorted(parent_options.iteritems(), key=lambda x: x[0]):
-                            info['options'][v] = k
-                    else:
-                        parent_options = getattr(parent_class, 'ENUM_' + field_name)
-                        for k in info['options'].iterkeys():
-                            info['options'][k] = parent_options[k]
-                except Exception as err:
-                    print err
-                    raise err
+                parent_class = collection.find_by_name(parent_name)
```
Comment (mluessi) [inline python/taulabs/uavo.py]: Nice :)

**Comment by mluessi** [comment]
Very Nice :).

File: python/get_config.py (hunk 1 above)
Comment (mlyle) [inline python/get_config.py]: This is in various of the command line tools so you can do python/dumplog.py or ~/taulabs/python/dumplog.py

Comment (mluessi) [inline python/get_config.py]: OK, it certainly doesn't hurt but shouldn't be needed. So I'm fine with leaving it there.

File: python/get_config.py (hunk 2 above)
Comment (mluessi) [inline python/get_config.py]: PS: sorry for nitpicking ;)

Comment (mlyle) [inline python/get_config.py]: No worries :) fixed.

File: python/get_config.py (hunk 1 above)
Comment (mlyle) [inline python/get_config.py]: Yah, it doesn't seem necessary, but something blew up when I tried to remove it awhile ago.  This is from before when I was involved.

**Comment by mluessi** [comment]
Looks good to be merged.

**Comment by kubark42** [comment]
Merging based on @mluessi's feedback.

**Pull request merged**